### DISCUSSION TITLE
Use duckdb to flatten a JSON

### DISCUSSION THREAD

**Discussion by aborruso** [duckdb/duckdb#6749] [Q&A] [answered]
Hi,
I have this kind of JSON:

```json
{
  "CAR": "203",
  "DES_TIPO_MISURA": "Regime di aiuti",
  "COD_CE_MISURA": "SA.40429",
  "COMPONENTI_AIUTO": {
    "COMPONENTE_AIUTO": [
      {
        "ID_COMPONENTE_AIUTO": "12132797",
        "COD_PROCEDIMENTO": "3"
      },
      {
        "ID_COMPONENTE_AIUTO": "13132797",
        "COD_PROCEDIMENTO": "5"
      }
    ]
  }
}
```

This is one of several records. I have this goal: to have in output the `COMPONENTI_AIUTO.COMPONENTE_AIUTO` subfields and add to these some superior hierarchical element.

| ID_COMPONENTE_AIUTO | COD_PROCEDIMENTO | CAR |
| --- | --- | --- |
| 12132797 | 3 | 203 |
| 13132797 | 5 | 203 |

I think I should use UNNEST funciotn, but I do not understand how to use it properly. If I run thish

```
select UNNEST(COMPONENTI_AIUTO.COMPONENTE_AIUTO),CAR from  read_json('sample.jsonl', json_format='records', auto_detect=true)
```

I have

| unnest("COMPONENTI_AIUTO"."COMPONENTE_AIUTO") | CAR |
| --- | --- |
| {'ID_COMPONENTE_AIUTO': 12132797, 'COD_PROCEDIMENTO': 3} | 203 |
| {'ID_COMPONENTE_AIUTO': 13132797, 'COD_PROCEDIMENTO': 5} | 203 |

How to make it properly?

Thank you

**Comment by Mytherin** [comment, marked as answer]
`UNNEST` removes one level of nesting from a list, so it turns `INT[]` into `INT`, or in your example `STRUCT(BIGINT,BIGINT)[]` into `STRUCT(BIGINT,BIGINT)`.

You can use the `.*` operator to expand all elements of a struct. Right now you are limited to doing this on top-level elements, however, so you need to use a subquery:

```sql
SELECT unnested.*, CAR FROM (
    SELECT UNNEST(COMPONENTI_AIUTO.COMPONENTE_AIUTO) AS unnested,CAR
    FROM read_json('sample.jsonl', json_format='records', auto_detect=true));
```

**Reply by aborruso** [reply]
duckdb is a gem: thank you!

**Reply by aborruso** [reply]
Dear @Mytherin ,
I'm trying to testing it, to another structure. I have a JSON field, not a struct one.

![image](https://user-images.githubusercontent.com/30607/225612505-9197d3ef-b992-4c5f-9fdb-54340e96d38c.png)


What's the general function to unnest it?

Thank you

**Reply by aborruso** [reply]
> I'm trying to testing it, to another structure. I have a JSON field, not a struct one.

I can use something like

```
SELECT json_extract_string(COMPONENTI_AIUTO.COMPONENTE_AIUTO,'$.ID_COMPONENTE_AIUTO') u,CAR FROM read_ndjson_auto('sample.jsonl', auto_detect=true);
```

Is there a way to use for the `path` some jolly, and not specify all field names? Something like `'$.*'`

Thank you

**Reply by Mytherin** [reply]
For `JSON` fields the fields are not known up-front and can vary within each row, as such you do need to specify each column individually. Have a look [in the JSON blog post](https://duckdb.org/2023/03/03/json.html#handling-inconsistent-json-schemas) for some tips on how to handle this type of data.

**Reply by aborruso** [reply]
Once again, thank you , you are always very kind
